core/blockarchiver/types.go

```diff
@@ -0,0 +1,169 @@
+	Jsonrpc string     `json:"jsonrpc,omitempty"`
+	Result  []*Block   `json:"result,omitempty"`
+}
+
+// GetBundleNameResponse represents a response from the getBundleName RPC call
+type GetBundleNameResponse struct {
+	Data string `json:"data"`
+}
+
+// Transaction represents a transaction in the Ethereum blockchain
+type Transaction struct {
+	BlockHash            string        `json:"blockHash"`
+	BlockNumber          string        `json:"blockNumber"`
+	From                 string        `json:"from"`
+	Gas                  string        `json:"gas"`
+	GasPrice             string        `json:"gasPrice"`
+	Hash                 string        `json:"hash"`
+	Input                string        `json:"input"`
+	Nonce                string        `json:"nonce"`
+	To                   string        `json:"to"`
+	TransactionIndex     string        `json:"transactionIndex"`
+	Value                string        `json:"value"`
+	Type                 string        `json:"type"`
+	AccessList           []AccessTuple `json:"accessList"`
+	ChainId              string        `json:"chainId"`
+	V                    string        `json:"v"`
+	R                    string        `json:"r"`
+	S                    string        `json:"s"`
+	YParity              string        `json:"yParity"`
+	MaxPriorityFeePerGas string        `json:"maxPriorityFeePerGas"`
+	MaxFeePerGas         string        `json:"maxFeePerGas"`
+	MaxFeePerDataGas     string        `json:"maxFeePerDataGas"`
+	MaxFeePerBlobGas     string        `json:"maxFeePerBlobGas"`
+	BlobVersionedHashes  []string      `json:"blobVersionedHashes"`
+}
+
+// AccessTuple represents a tuple of an address and a list of storage keys
+type AccessTuple struct {
+	Address     string
+	StorageKeys []string
+}
+
+// GeneralBlock represents a block in the Ethereum blockchain
+type GeneralBlock struct {
+	*types.Block
+	TotalDifficulty *big.Int `json:"totalDifficulty"` // Total difficulty in the canonical chain up to and including this block.
+}
+
+// Range represents a range of Block numbers
+type Range struct {
+	from uint64
+	to   uint64
+}
+
+// RequestLock is a lock for making sure we don't fetch the same bundle concurrently
+type RequestLock struct {
+	rangeMap  map[uint64]Range
+	lookupMap map[uint64]bool
+	mu        sync.RWMutex
+}
+
+// NewRequestLock creates a new RequestLock
+func NewRequestLock() *RequestLock {
+	return &RequestLock{
+		rangeMap:  make(map[uint64]Range),
+		lookupMap: make(map[uint64]bool),
+	}
+}
+
+// IsWithinAnyRange checks if the number is within any of the cached ranges
+func (rc *RequestLock) IsWithinAnyRange(num uint64) bool {
+	rc.mu.RLock()
+	defer rc.mu.RUnlock()
+	_, exists := rc.lookupMap[num]
+	return exists
+}
+
+// AddRange adds a new range to the cache
+func (rc *RequestLock) AddRange(from, to uint64) {
+	rc.mu.Lock()
+	defer rc.mu.Unlock()
+	rc.rangeMap[from] = Range{from, to}
+	for i := from; i <= to; i++ {
+		rc.lookupMap[i] = true
```

Comment:
This maybe not efficient.. anyway to improve?

Reply:
added todo for now